### PR TITLE
refact: Add db.DB wrapper and refact all test fixtures to use it.

### PR DESCRIPTION
**Pull request by jimlambrt**
This PR will allow us to stop leaking the gorm abstraction within
our unit tests.

This PR is a fast-follower to https://github.com/hashicorp/boundary/pull/1528. It cannot be merged until that PR is accepted and then it should be rebased against main.

### CODE REVIEW

File: internal/cmd/base/servers.go [hunk 1]
```diff
@@ -575,7 +575,7 @@ func (b *Server) ConnectToDatabase(dialect string) error {
 				Colorful: false,        // Disable color
 			},
 		)
-		dbase = dbase.Session(&gorm.Session{Logger: newLogger})
+		dbase.DB = dbase.Session(&gorm.Session{Logger: newLogger})
```
Comment (tmessi) [inline internal/cmd/base/servers.go:578]: This is still causing `gorm` and `gorm/logger` to be imported here. I think that can be elimnated by moving this logger configuration into the `db` package. Probably by adding some `Options` to `db`, i.e.:

```go
opts := []db.Option
if os.Getenv("BOUNDARY_DISABLE_GORM_FORMATTER") == "" {
  opts = append(opts, db.WithGormFormatter(b.Logger))
}
dbase, err := db.Open(dbType, b.DatabaseUrl, opts...)
```

then in `db`:
```go
package db

// i dont think i have this exactly correct, but hopefully its close enough to get the idea
func Open(dbType DbType, connectionUrl string, options ...Option) (*DB, error) {
  // ...
  if opts.WithGormFormatter != nil {
    newLogger := logger.New(
      db.GetGormLogger(opts.WithGormFormatter),
      logger.Config{
        LogLevel: logger.Error,
        Colorful: false,
      },
    )
    db.DB = db.Session(&gorm.Session{Logger: newLogger})
  }

  // ...
}

File: internal/cmd/base/servers.go [hunk 2]
```diff
@@ -560,7 +560,7 @@ func (b *Server) ConnectToDatabase(dialect string) error {
 	if b.DatabaseMaxOpenConnections == 1 {
 		return fmt.Errorf("unable to create db object with dialect %s: %s", dialect, "max_open_connections must be unlimited by setting 0 or at least 2")
 	} else {
-		underlyingDB, err := dbase.DB()
+		underlyingDB, err := dbase.SqlDB()
```
Comment (tmessi) [inline internal/cmd/base/servers.go:563]: likewise, the setting of MaxOpenConnections could be moved into an Option for the call to `db.Open`

File: internal/cmd/base/servers.go [hunk 3]
```diff
@@ -620,7 +620,7 @@ func (b *Server) CreateGlobalKmsKeys(ctx context.Context) error {
 
 func (b *Server) DestroyDevDatabase() error {
 	if b.Database != nil {
-		underlyingDB, err := b.Database.DB()
+		underlyingDB, err := b.Database.SqlDB()
 		if err != nil {
 			return err
 		}
```
Comment (tmessi) [inline internal/cmd/base/servers.go:626]: I think this `Close` logic should probably be moved into a method on `db.DB`, ie:

```go
func (d *DB) Close() error {
  underlyingDB, err := d.SqlDB()
  if err != nil {
    return err
  }
  return underlying.Close()
}
```
}

File: internal/db/db.go [hunk 4]
```diff
@@ -34,9 +35,23 @@ func StringToDbType(dialect string) (DbType, error) {
...
 }
 
+// DB is a wrapper around the ORM
+type DB struct {
+	*gorm.DB
+}
+
+// SqlDB returns the underlying driver.
+//
+// Note: This func is not named DB(), because our chosen of ORM (gorm) which is
```
Comment (tmessi) [inline internal/db/db.go:45]: nitpick: `s/chosen/choice/`

File: internal/db/db.go [hunk 5]
```diff
@@ -34,9 +35,23 @@ func StringToDbType(dialect string) (DbType, error) {
...
+// embedded for various reasons, already exports a DB() func and it's base type
+// is also an export DB type
+func (d *DB) SqlDB() (*sql.DB, error) {
+	return d.DB.DB()
```
Comment (tmessi) [inline internal/db/db.go:49]: I think this should check if `d.DB` is `nil` and return an error if it is.

File: internal/db/db.go (hunk 4 above)
Comment (jimlambrt) [reply internal/db/db.go:45]: ty!

File: internal/db/db.go [hunk 6]
```diff
@@ -34,9 +35,23 @@ func StringToDbType(dialect string) (DbType, error) {
 	}
 }
 
+// DB is a wrapper around the ORM
+type DB struct {
+	*gorm.DB
+}
+
+// SqlDB returns the underlying driver.
```
Comment (tmessi) [inline internal/db/db.go:43]: nitpick: I think technically this isn't the "driver" its the `sql.DB` which its [docs](https://pkg.go.dev/database/sql#DB) describe as:

> DB is a database handle representing a pool of zero or more underlying connections.

The drivers are [these packages](https://github.com/golang/go/wiki/SQLDrivers)

File: internal/db/db.go (hunk 5 above)
Comment (jimlambrt) [reply internal/db/db.go:49]: ty!

File: internal/cmd/base/servers.go (hunk 3 above)
Comment (jimlambrt) [reply internal/cmd/base/servers.go:626]: ty!

File: internal/db/db.go [hunk 7]
```diff
@@ -34,9 +37,42 @@ func StringToDbType(dialect string) (DbType, error) {
...
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	return underlying.Close()
+}
+
+// Open a database connection which is long-lived. The options of
+// WithGormFormatter and WithMaxOpenConnections are supported.
+//
+// Note: you need to call Close() on the returned DB
```
Comment (tmessi) [inline internal/db/db.go:74]: interestingly, the `sql` docs for `Close` state:

> It is rare to Close a DB, as the DB handle is meant to be long-lived and shared between many goroutines.

And for `Open`:

> It is rarely necessary to close a DB.

And I think the only spot we are calling Close (outside of tests) in is `DestroyDevDatabase` to ensure the connections are closed before destroying the database.

Comment (jimlambrt) [reply internal/db/db.go]: updated comment/doc

**Review by tmessi — APPROVED** [review]

**Review by tmessi — APPROVED** [review]

**Pull request merged**